nifi-extension-bundles/nifi-aws-bundle/nifi-aws-kinesis/src/main/java/org/apache/nifi/processors/aws/kinesis/MemoryBoundRecordBuffer.java

```diff
@@ -0,0 +1,673 @@
+        final ShardBuffer buffer = shardBuffers.get(bufferId);
+        if (buffer == null) {
+            logger.debug("Buffer with id {} not found. Cannot checkpoint the ended shard", bufferId);
+            return;
+        }
+
+        logger.debug("Finishing consumption for buffer {}. Checkpointing the ended shard", bufferId);
+        buffer.checkpointEndedShard(checkpointer);
+
+        logger.debug("Removing buffer with id {} after successful ended shard checkpoint", bufferId);
+        shardBuffers.remove(bufferId);
+    }
+
+    @Override
+    public void shutdownShardConsumption(final ShardBufferId bufferId, final RecordProcessorCheckpointer checkpointer) {
+        final ShardBuffer buffer = shardBuffers.get(bufferId);
+        if (buffer == null) {
+            logger.debug("Buffer with id {} not found. Cannot shutdown shard consumption", bufferId);
+            return;
+        }
+
+        logger.debug("Shutting down the buffer {}. Checkpointing last consumed record", bufferId);
+        buffer.shutdownBuffer(checkpointer);
+
+        logger.debug("Removing buffer with id {} after successful last consumed record checkpoint", bufferId);
+        shardBuffers.remove(bufferId);
+    }
+
+    @Override
+    public void consumerLeaseLost(final ShardBufferId bufferId) {
+        final ShardBuffer buffer = shardBuffers.remove(bufferId);
+
+        logger.debug("Lease lost for buffer {}: Invalidating", bufferId);
+
+        if (buffer != null) {
+            final Collection<RecordBatch> invalidatedBatches = buffer.invalidate();
+            memoryTracker.freeMemory(invalidatedBatches);
+        }
+    }
+
+    @Override
+    public Optional<ShardBufferLease> acquireBufferLease() {
+        final Set<ShardBufferId> seenBuffers = new HashSet<>();
+
+        while (true) {
+            final ShardBufferId bufferId = buffersToLease.poll();
+            if (bufferId == null) {
+                // The queue is empty or all buffers were seen already. Nothing to consume.
+                return Optional.empty();
+            }
+
+            if (seenBuffers.contains(bufferId)) {
+                // If the same buffer is seen again, there is a high chance we iterated through most of the buffers and didn't find any that isn't empty.
+                // To avoid burning CPU we return empty here, even if some buffer received records in the meantime. It will be picked up in the next iteration.
+                buffersToLease.add(bufferId);
+                return Optional.empty();
+            }
+
+            final ShardBuffer buffer = shardBuffers.get(bufferId);
+
+            if (buffer == null) {
+                // By the time the bufferId is polled, it might have been invalidated. No need to return it to the queue.
+                logger.debug("Buffer with id {} was removed while polling for lease. Continuing to poll.", bufferId);
```

Comment:
Minor detail, but for consistency, log messages should not end with a `.` character.
```suggestion
                logger.debug("Buffer with id {} was removed while polling for lease. Continuing to poll", bufferId);
```